src/ray/gcs/redis_context.cc

```diff
@@ -203,25 +204,37 @@ void GlobalRedisCallback(void *c, void *r, void *privdata) {
   ProcessCallback(callback_index, std::make_shared<CallbackReply>(reply));
 }
 
-int64_t RedisCallbackManager::add(const RedisCallback &function, bool is_subscription,
-                                  boost::asio::io_service &io_service) {
+int64_t RedisCallbackManager::AllocateCallbackIndex() {
+  std::lock_guard<std::mutex> lock(mutex_);
+  return num_callbacks_++;
+}
+
+int64_t RedisCallbackManager::AddCallback(const RedisCallback &function,
+                                          bool is_subscription,
+                                          boost::asio::io_service &io_service,
+                                          int64_t callback_index) {
   auto start_time = absl::GetCurrentTimeNanos() / 1000;
 
   std::lock_guard<std::mutex> lock(mutex_);
+  if (callback_index == -1) {
```

Comment:
num_callbacks_ is start from 0, why callback_index will be -1, it maybe seems confusing.

Reply:
-1 means that the callback index is not set.